jablib/src/main/java/org/jabref/logic/util/io/BrokenLinkedFileNameBasedFileFinder.java

```diff
@@ -0,0 +1,60 @@
+package org.jabref.logic.util.io;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.jabref.model.entry.BibEntry;
+
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class BrokenLinkedFileNameBasedFileFinder implements FileFinder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BrokenLinkedFileNameBasedFileFinder.class);
+
+    @Override
+    public List<Path> findAssociatedFiles(@NonNull BibEntry entry,
+                                          @NonNull List<Path> directories,
+                                          List<String> extensions) throws IOException {
+        Set<String> brokenLinkedFileNames = getBrokenLinkedFileNames(entry, directories);
+
+        if (brokenLinkedFileNames.isEmpty()) {
+            return List.of();
+        }
+
+        List<Path> matches = new ArrayList<>();
+        for (Path directory : directories) {
+            matches.addAll(findAssociatedFilesInDirectory(directory, brokenLinkedFileNames, extensions));
+        }
+        return matches;
+    }
+
+    private static List<Path> findAssociatedFilesInDirectory(
+            Path directory,
+            Set<String> brokenLinkedFileNames,
+            List<String> extensions) throws IOException {
+        try (Stream<Path> walk = Files.walk(directory)) {
+            // find all files (not directory) that match:
+            // 1. it has the same name (case-insensitive) as one of the broken linked file names
+            // 2. its file extension matches one specified in extensions
```

Comment:
Here we do not consider a strict file extension match as:
- `Entry Editor - Automatically search and show unlinked files in the entity editor` may want to show those file with same name but different extension as possible candidate
- `Quality - Automatically set file links` will make sure only file with the same name and extention is auto linked. The logic is in `autoLinkBrokenLinkedFiles`